base/traits.jl

```diff
@@ -20,3 +20,40 @@ TypeArithmetic(instance) = TypeArithmetic(typeof(instance))
 TypeArithmetic(::Type{<:AbstractFloat}) = ArithmeticRounds()
 TypeArithmetic(::Type{<:Integer}) = ArithmeticOverflows()
 TypeArithmetic(::Type{<:Any}) = ArithmeticUnknown()
+
+# trait for objects that support ranges with regular step
+"""
+    TypeRangeStep(instance)
+    TypeRangeStep(T::Type)
+
+Indicate whether an instance or a type supports constructing a range with
+a perfectly regular step or not. A regular step means that
+[`step`](@ref) will always be exaclty equal to the difference between two
+subsequent elements in a range, i.e. for a range `r::Range{T}`:
+```julia
+all(diff(r) .== step(r))
+```
+
+When a type `T` always leads to ranges with regular steps, it should
+define the following method:
+```julia
+Base.TypeRangeStep(::Type{<:Range{<:T}}) = RangeStepRegular()
+```
+This will allow [`hash`](@ref) to use an O(1) algorithm for `Range{T}` objects
+instead of the default O(N) algorithm (with N the length of the range).
+
+In some cases, whether the step will be regular depends not only on the
+element type `T`, but also on the type of the step `S`. In that case, more
+specific methods should be defined:
+```julia
+Base.TypeRangeStep(::Type{<:OrdinalRange{<:T, <:S}}) = RangeStepRegular()
+```
+
+By default, all range types are assumed to be `RangeStepIrregular`, except
+subtypes of `Integer`.
+"""
+abstract type TypeRangeStep end
+struct RangeStepRegular   <: TypeRangeStep end # range with regular step
+struct RangeStepIrregular <: TypeRangeStep end # range with rounding error
+
+TypeRangeStep(instance) = TypeRangeStep(typeof(instance))
```

Comment:
The only situation in which package authors would need this would be if they define a new range type for which neither the O(N) `AbstractRange` fallback nor the O(1) `hash_range` method works, and yet the custom type's method would need to check whether a given type uses a regular step. I can't find an example where this would apply, especially given that a custom `hash` implementation for a range should be consistent with other `AbstractArray` implementations.